nes-network/network/src/sender/channel.rs

```diff
@@ -0,0 +1,387 @@
+            return Err(ErrorOrStatus::Status(
+                ChannelHandlerStatus::ClosedBySoftwareButFailedToPropagate(
+                    "Did not try to propagate".into(),
+                ),
+            ));
+        };
+        Ok(r)
+    }
+
+    async fn read_from_other_side(
+        cancellation_token: &CancellationToken,
+        s: &mut DataChannelSenderReader<R>,
+    ) -> InternalResult<DataChannelResponse> {
+        let Some(r) = Self::cancelable(cancellation_token, s.next()).await? else {
+            return Err(ErrorOrStatus::Error("Connection Lost".into()));
+        };
+
+        match r {
+            Err(e) => Err(ErrorOrStatus::Error(e.into())),
+            Ok(r) => Ok(r),
+        }
+    }
+
+    async fn run_internal(&mut self) -> core::result::Result<(), ErrorOrStatus> {
+        loop {
+            if self.pending_writes.is_empty() || self.wait_for_ack.len() >= MAX_PENDING_ACKS {
+                select! {
+                    response = Self::read_from_other_side(&self.cancellation_token, &mut self.reader) => self.handle_response(response?)?,
+                    request = Self::read_from_software(&self.cancellation_token, &mut self.queue) => self.handle_request(request?).await?,
+                }
+            } else {
+                select! {
+                    response = Self::read_from_other_side(&self.cancellation_token, &mut self.reader) => self.handle_response(response?)?,
+                    request = Self::read_from_software(&self.cancellation_token, &mut self.queue) => self.handle_request(request?).await?,
+                    send_result = Self::send_pending(&self.cancellation_token, &mut self.writer, &mut self.pending_writes, &mut self.wait_for_ack) => send_result?,
+                }
+            }
+        }
+    }
+    async fn run(&mut self) -> Result<ChannelHandlerStatus> {
+        let status = match self
+            .run_internal()
+            .await
+            .expect_err("Handler should have not terminated by its own.")
+        {
+            ErrorOrStatus::Error(e) => return Err(e),
+            ErrorOrStatus::Status(status) => status,
+        };
+
+        // Fixup the ClosedBySoftwareButFailedToPropagate created in read_from_software
+        if let ChannelHandlerStatus::ClosedBySoftwareButFailedToPropagate(_) = status {
+            let Some(result) = self
+                .cancellation_token
+                .run_until_cancelled(self.writer.send(DataChannelRequest::Close))
+                .await
+            else {
+                return Ok(ChannelHandlerStatus::Cancelled);
+            };
+            if let Err(e) = result {
+                return Ok(ChannelHandlerStatus::ClosedBySoftwareButFailedToPropagate(
+                    e.into(),
+                ));
+            };
+            return Ok(ChannelHandlerStatus::ClosedBySoftware);
+        }
+
+        Ok(status)
+    }
+}
+
+async fn channel_handler(
+    cancellation_token: CancellationToken,
```

Comment:
The channel_id, cancellation pair and the two connection ids could be combined somehow, there are always passed together down the stack. 

Also, communication could somehow be hidden in another type, it is passed numerous times but used only at the lowest level of abstraction (actually transmitting messages across the channel).